nacos/src/main/java/com/linecorp/armeria/internal/nacos/LoginClient.java

```diff
@@ -122,6 +91,39 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
         return HttpResponse.of(future);
     }
 
+    private final class CachedLoginResult {
+        @Nullable
+        private volatile CacheEntry cacheEntry;
+
+        CachedLoginResult() { }
+
+        CompletableFuture<String> get() {
+            final CacheEntry cacheEntry = this.cacheEntry;
+            if (cacheEntry != null && !cacheEntry.isExpired()) {
+                return UnmodifiableFuture.completedFuture(cacheEntry.loginResult.accessToken);
+            }
+
+            return login().thenApply(loginResult -> {
```

Comment:
This may result in sending multiple login requests that are not what we really want.
If `login()` is being performed, other requests can piggyback on the returned future.

By the way, I think an `AsyncLoader` loading a value when it is expired or periodically would be useful in many places.
Created https://github.com/line/armeria/issues/5506

Reply:
Maybe we shouldn't let @KonaEspresso94 wait for it to be merged too long and use Caffeine internally and later replace the Caffeine usage with `AsyncLoader`?

Reply:
I didn't know it would take this long. 😅 Let's leave that part as is. I will continue to review this PR tomorrow.

Reply:
Thank you for your consideration. Following your comments, I have reverted the commit to use Caffeine instead of my own implementation.